### PR TITLE
test: slashing integration todos

### PR DESCRIPTION
**Pull request by 0xClandestine**
### Changes:

- adds `HOLDS_MAX` assetType (denotes a user that holds the maximum allowed strategies to test gas constraints).
- adds `print.gasUsed()` (prints the gas consumed from the last call).
- adds gas logging via `print.gasUsed()`.
- adds `PermissionController` support to `AVS` and `User` integration user types.
- removes **most** `assertApproxEqAbs` (1 single instance remains for beaconchain).

### TODOS:
- add remaining slashing related checks.
- add remaining slashing scenarios.

### CODE REVIEW

File: src/test/integration/IntegrationBase.t.sol [hunk 1]
```diff
@@ -14,7 +14,7 @@ import "src/test/integration/TimeMachine.t.sol";
 import "src/test/integration/users/User.t.sol";
 import "src/test/integration/users/User_M1.t.sol";
 
-abstract contract IntegrationBase is IntegrationDeployer {
+abstract contract IntegrationBase is IntegrationDeployer, IAllocationManagerTypes {
```
Comment (8sunyuan) [inline src/test/integration/IntegrationBase.t.sol:17]: Can we remove IAllocationManagerTypes

File: src/test/integration/IntegrationBase.t.sol [hunk 2]
```diff
@@ -351,6 +347,234 @@ abstract contract IntegrationBase is IntegrationDeployer {
...
+                );
+            }
+
+            // Check magnitudes
+            Magnitudes memory curMagnitude = curMagnitudes[i];
+            Magnitudes memory prevMagnitude = prevMagnitudes[i];
+
+            // if (isAllocation) ...
+            if (prevAlloc.currentMagnitude < allocateParams.newMagnitudes[i]) {
+                uint256 delta = allocateParams.newMagnitudes[i] - prevAlloc.currentMagnitude;
```
Comment (8sunyuan) [inline src/test/integration/IntegrationBase.t.sol:401]: Can we move this delta calculation above the if statement and then do the check. Would be a little easier to read 
and its being done in both if else blocks
```
// if (isAllocation)
if (delta > 0) {
...

// if (deallocation)
} else {
}

File: src/test/integration/IntegrationBase.t.sol [hunk 3]
```diff
@@ -351,6 +347,234 @@ abstract contract IntegrationBase is IntegrationDeployer {
...
+                
+                if (completed) {
+                    // FIXME: fix completed case
```
Comment (8sunyuan) [inline src/test/integration/IntegrationBase.t.sol:404]: If its an allocation, doesn't it add to encumbered magnitude(and therefore reducing allocatable by same amount)
regardless if the delay has passed? So we can remove the if check and the encumbered/allocatable values are always updated regardless if completed or not
https://github.com/Layr-Labs/eigenlayer-contracts/blob/d3109212d8869ebba791790bfe0e22bdfadd7e5f/src/contracts/core/AllocationManager.sol#L206-L211

File: src/test/integration/IntegrationBase.t.sol [hunk 4]
```diff
@@ -351,6 +347,234 @@ abstract contract IntegrationBase is IntegrationDeployer {
...
+
+            uint64 slashedMagnitude = uint64(uint256(prevAlloc.currentMagnitude).mulWadRoundUp(slashingParams.wadsToSlash[i]));
+
+            assertEq(
+                curAlloc.currentMagnitude, 
+                prevAlloc.currentMagnitude - slashedMagnitude, 
+                string.concat(err, " (currentMagnitude)")
+            );
+
+            if (completed) {
```
Comment (8sunyuan) [inline src/test/integration/IntegrationBase.t.sol:496]: Even if minimal, very much appreciate the comments done in 
`assert_Snap_Allocations_Modified` which indicate what block of code is checking which values like 
`// Max magnitude should not have mutated.`, `// Check allocations`, etc

File: src/test/integration/IntegrationBase.t.sol [hunk 5]
```diff
@@ -351,6 +347,234 @@ abstract contract IntegrationBase is IntegrationDeployer {
...
+                prevMagnitude.max, 
+                string.concat(err, " (max magnitude)")
+            );
+        }
+    }
+
+    function assert_Snap_Allocations_Slashed(
+        SlashingParams memory slashingParams,
+        OperatorSet memory operatorSet,
+        bool completed,
```
Comment (8sunyuan) [inline src/test/integration/IntegrationBase.t.sol:472]: What does completed mean in this context? When would this value be false?

Comment (0xClandestine) [reply src/test/integration/IntegrationBase.t.sol:472]: It's supposed to differentiate whether the allocation was completed before slashed or not.

File: src/test/integration/IntegrationBase.t.sol [hunk 6]
```diff
@@ -351,6 +347,234 @@ abstract contract IntegrationBase is IntegrationDeployer {
...
+                        uint64(uint256(uint128(-prevAlloc.pendingDiff)).mulWadRoundUp(slashingParams.wadsToSlash[i]));
+                    
+                    assertEq(
+                        curAlloc.pendingDiff, 
+                        prevAlloc.pendingDiff + int128(uint128(slashedPending)), 
+                        string.concat(err, " (pendingDiff)")
+                    );
+                }
+
+                assertEq(
```
Comment (8sunyuan) [inline src/test/integration/IntegrationBase.t.sol:517]: I think the effectBlock here can vary depending on if we have a pending allocation or deallocation. Need to check for that

File: src/contracts/pods/EigenPod.sol [hunk 7]
```diff
@@ -397,8 +397,8 @@ contract EigenPod is Initializable, ReentrancyGuardUpgradeable, EigenPodPausingC
      * @dev Reverts if `amountWei` is not a whole Gwei amount
      */
     function withdrawRestakedBeaconChainETH(address recipient, uint256 amountWei) external onlyEigenPodManager {
-        require(amountWei % GWEI_TO_WEI == 0, AmountMustBeMultipleOfGwei());
```
Comment (ypatil12) [inline src/contracts/pods/EigenPod.sol:400]: Can we get a unit test on this

**Review by ypatil12 — CHANGES_REQUESTED** [review]
Need unit test on EP change

File: src/test/integration/IntegrationBase.t.sol [hunk 8]
```diff
@@ -351,6 +347,234 @@ abstract contract IntegrationBase is IntegrationDeployer {
...
+
+            assertEq(
+                curMagnitude.max, 
+                prevMagnitude.max - slashedMagnitude, 
+                string.concat(err, " (max magnitude)")
+            );
+        }
+    }
+
+    function assert_HasUnderlyingTokenBalances_AfterSlash(
```
Comment (8sunyuan) [inline src/test/integration/IntegrationBase.t.sol:547]: For non native ETH Strategies, we should also have an assertion on the token balance of the Strategy being burned/updated after the `slashOperator` call and it should be sent to the burn address.
Perhaps it makes sense to be in this check or a separate assert function

File: src/test/integration/tests/Deposit_Delegate_Allocate.t.sol [hunk 9]
```diff
@@ -27,21 +30,161 @@ contract Integration_Deposit_Delegate_Allocate is IntegrationCheckUtils {
...
+        for (uint256 i; i < operatorSets.length; ++i) {
+            assert_Snap_Allocations_Modified(
+                operator, allocateParams[i], true, "operator allocations should be updated after delay"
+            );
+        }
     }
-}
\ No newline at end of file
+
+    function testFuzz_deposit_delegate_allocate_slash_undelegate_completeAsTokens(
```
Comment (8sunyuan) [inline src/test/integration/tests/Deposit_Delegate_Allocate.t.sol:62]: To clarify, this is and the other slash + completeAsTokens integration tests failed due to the EigenPod gwei bug correct?

File: src/test/integration/tests/Deposit_Delegate_Allocate.t.sol [hunk 10]
```diff
@@ -27,21 +30,161 @@ contract Integration_Deposit_Delegate_Allocate is IntegrationCheckUtils {
...
+        // 5. Undelegate from an operator
+        IDelegationManagerTypes.Withdrawal[] memory withdrawals = staker.undelegate();
+        bytes32[] memory withdrawalRoots = _getWithdrawalHashes(withdrawals);
+        // 6. Complete withdrawal
+        _rollBlocksForCompleteWithdrawals();
+        for (uint256 i = 0; i < withdrawals.length; ++i) {
+            uint256[] memory expectedTokens =
+                _calculateExpectedTokens(withdrawals[i].strategies, withdrawals[i].scaledShares);
+            staker.completeWithdrawalAsTokens(withdrawals[i]);
+            // FIXME: check_Withdrawal_AsTokens_State_AfterSlash(staker, operator, withdrawals[i], allocateParams, slashingParams, expectedTokens);
```
Comment (8sunyuan) [inline src/test/integration/tests/Deposit_Delegate_Allocate.t.sol:112]: Whats broken in this check atm?

Comment (0xClandestine) [reply src/test/integration/tests/Deposit_Delegate_Allocate.t.sol:112]: currently failing for staker shares, which should also be conditionally adjusted

File: src/test/integration/tests/Deposit_Delegate_Allocate.t.sol (hunk 9 above)
Comment (0xClandestine) [reply src/test/integration/tests/Deposit_Delegate_Allocate.t.sol:62]: yes

File: src/test/integration/tests/Deposit_Delegate_Allocate.t.sol [hunk 11]
```diff
@@ -4,21 +4,24 @@ pragma solidity ^0.8.27;
 import "src/test/integration/IntegrationChecks.t.sol";
 import "src/test/integration/users/User.t.sol";
 
-contract Integration_Deposit_Delegate_Allocate is IntegrationCheckUtils {
-    function testFuzz_deposit_delegate_allocate(uint24 _random) public {
+// TODO: move randomness from tests
+
+contract Integration_Deposit_Delegate_Allocate is IntegrationCheckUtils, IDelegationManagerTypes {
```
Comment (8sunyuan) [inline src/test/integration/tests/Deposit_Delegate_Allocate.t.sol:9]: Same thing here and the other test file, could we remove the IDelegationManagerTypes inheritance?

File: src/contracts/pods/EigenPod.sol (hunk 7 above)
Comment (0xClandestine) [reply src/contracts/pods/EigenPod.sol:400]: Already done `testFuzz_withdrawRestakedBeaconChainETH_AmountGweiNotDivisibleByGwei` in `EigenPodUnit.t.sol`.

File: src/test/integration/IntegrationBase.t.sol (hunk 8 above)
Comment (0xClandestine) [reply src/test/integration/IntegrationBase.t.sol:537]: will include in follow up pr

File: src/test/integration/tests/Deposit_Delegate_Allocate.t.sol (hunk 10 above)
Comment (0xClandestine) [reply src/test/integration/tests/Deposit_Delegate_Allocate.t.sol:113]: will fix in follow up pr

File: src/test/integration/IntegrationBase.t.sol (hunk 1 above)
Comment (0xClandestine) [reply src/test/integration/IntegrationBase.t.sol]: fixed in [d7a5d61](https://github.com/Layr-Labs/eigenlayer-contracts/pull/934/commits/d7a5d61437737a470dd4d22220e8a818d21c2ac1)

File: src/test/integration/IntegrationBase.t.sol (hunk 3 above)
Comment (0xClandestine) [reply src/test/integration/IntegrationBase.t.sol]: fixed in [d7a5d61](https://github.com/Layr-Labs/eigenlayer-contracts/pull/934/commits/d7a5d61437737a470dd4d22220e8a818d21c2ac1)

File: src/test/integration/IntegrationBase.t.sol (hunk 6 above)
Comment (0xClandestine) [reply src/test/integration/IntegrationBase.t.sol:506]: fixed in [d7a5d61](https://github.com/Layr-Labs/eigenlayer-contracts/pull/934/commits/d7a5d61437737a470dd4d22220e8a818d21c2ac1)

File: src/test/integration/tests/Deposit_Delegate_Allocate.t.sol [hunk 12]
```diff
@@ -27,21 +30,161 @@ contract Integration_Deposit_Delegate_Allocate is IntegrationCheckUtils {
...
+
+        AllocateParams[] memory allocateParams = new AllocateParams[](operatorSets.length);
+
+        for (uint256 i; i < operatorSets.length; ++i) {
             uint256 len = allocationManager.getStrategiesInOperatorSet(operatorSets[i]).length;
-            operator.modifyAllocations(operatorSets[i], _randMagnitudes({ sum: 1 ether / uint64(operatorSets.length), len: len }));
-            (IStrategy[] memory strategiesToSlash, uint[] memory wadsToSlash) = _randStrategiesAndWadsToSlash(operatorSets[i]);
-            avs.slashOperator(operator, operatorSets[i].id, strategiesToSlash, wadsToSlash);
+            allocateParams[i] = operator.modifyAllocations(
+                operatorSets[i], _randMagnitudes({sum: 1 ether / uint64(operatorSets.length), len: len})
```
[the remark this answers is not in the document]
Comment (0xClandestine) [reply src/test/integration/tests/Deposit_Delegate_Allocate.t.sol]: fixed in [d7a5d61](https://github.com/Layr-Labs/eigenlayer-contracts/pull/934/commits/d7a5d61437737a470dd4d22220e8a818d21c2ac1)

File: src/test/integration/tests/Deposit_Delegate_Allocate.t.sol (hunk 11 above)
Comment (0xClandestine) [reply src/test/integration/tests/Deposit_Delegate_Allocate.t.sol]: fixed in [d7a5d61](https://github.com/Layr-Labs/eigenlayer-contracts/pull/934/commits/d7a5d61437737a470dd4d22220e8a818d21c2ac1)

**Comment by ypatil12** [comment]
Closing in favor of https://github.com/Layr-Labs/eigenlayer-contracts/pull/945

**Pull request closed without merging**